Migrate CiviGrant component to an extension

colemanw opened the pull request:
Overview
----------------------------------------
This a straight port of CiviGrant to be packaged as an Extension instead of a Component. All the functionality is the same aside from how it's enabled or disabled in the UI.

Before
------------
Enable or disable CiviGrant through **Administer -> System Settings -> Components**. As a component it cannot be uninstalled.

After
------------
Enable, disable or uninstall CiviGrant through **Administer -> System Settings -> Extensions**. As an extension it can be uninstalled, which completely removes grant data from the database.

For existing sites, the extension will be automatically installed as either "enabled" or "disabled" depending on the status of the component.

For new sites, the extension is not installed by default.

colemanw committed:
Generate CiviGrant extension

colemanw committed:
CiviGrant - Move CRM & API code from core to extension

colemanw committed:
CiviGrant - Migrate component functions to hooks

colemanw committed:
CiviGrant - Migrate schema files

Insert and then delete CiviGrant to avoid messing up legacy code with expectations of component IDs.

colemanw committed:
CiviGrant - Move report files

colemanw committed:
CiviGrant - Install extension + managed entities during upgrade

Disable & delete component, install extension as either "enabled" or "disabled" depending on the status of the component.
Migrate option values to managed entities
Migrate navigation menu to extension managed entities

seamuslee001 commented:
mysql: [Warning] Using a password on the command line interface can be insecure.
ERROR 1146 (42S02) at line 3795: Table 'core220642fa2jcivi_z4j89.civicrm_grant' doesn't exist

colemanw committed:
CiviGrant - Move unit tests

colemanw committed:
CiviGrant - Remove or mitigate hardcoded references to civicrm_grant table in core

Removes CiviGrant from hardcoded lists when possible, or at least puts up
guards to check that it's available before attempting to use it.

colemanw committed:
CiviGrant - Migrate legacy search query code

This fixes up the `CRM_Contact_BAO_Query_Interface` class which had incorrectly
omitted `static` from some static functions.

Adds defaultReurnProperties to search hook interface.

JoeMurray commented:
@monishdeb could you do a review of this. Please include how this impacts extensions like Grant Application Pages, Grant Programs. I think there may be a profile edit one, and at least one financial regarding Sage Intacct. What is of interest and unlikely to be found is breakage of related extensions when CiviGrant is changed from Component to Extension. I'd focus on calls and installs. If related extensions need to check for enabled status of CiviGrant as an extension now that is fine.

monishdeb commented:
@colemanw when I tried to install the civigrant extension on a clean setup I got this error: 

<img width="1438" alt="Screenshot 2021-11-25 at 5 41 38 PM" src="https://user-images.githubusercontent.com/3735621/143439591-6229ca43-cc49-4ee3-af47-c3811b79e39d.png">

```
Nov 25 12:06:33  [info] $ManagedEntities_failed = Array
(
    [entity] => OptionValue
    [action] => create
    [params] => Array
        (
            [version] => 4
            [values] => Array
                (
                    [option_group_id.name] => grant_status
                    [label] => Submitted
                    [value] => 1
                    [name] => Submitted
                    [grouping] => 
                    [filter] => 0
                    [is_default] => 1
                    [weight] => 1
                    [description] => 
                    [is_optgroup] => 
                    [is_reserved] => 
                    [is_active] => 1
                    [icon] => 
                    [color] => 
                    [component_id] => 
                    [domain_id] => 
                    [visibility_id] => 
                )

            [checkPermissions] => 
        )

    [result] => Array
        (
            [fields] => Array
                (
                    [0] => option_group_id
                )

            [error_code] => mandatory_missing
            [entity] => OptionValue
            [action] => create
            [is_error] => 1
            [error_message] => Mandatory values missing from Api4 OptionValue::create: option_group_id
        )

)


```

colemanw commented:
@monishdeb I just tried it myself and was able to run it without error. I ran setup to generate a fresh database based on this branch to avoid conflicts:
```bash
git checkout civigrant
./bin/setup.sh
cv en civigrant
```

I also just made some additional fixes and force-pushed this branch. 3 things are still not working as they should:
1. Advanced search / Find grants
2. The contact summary tab (which is based on the same search code)
3. Reports

monishdeb commented on ext/civigrant/managed/OptionGroup_advanced_search_options_OptionValue_CiviGrant.mgd.php:
```diff
@@ -0,0 +1,31 @@
+    'update' => 'unmodified',
+    'params' => [
+      'version' => 4,
+      'values' => [
```
@colemanw shouldn't the parameters decalred directly under `params` not under `values` ?

colemanw replied:
@monishdeb - no because APIv4 structures it this way.

monishdeb commented:
@colemanw I have rebased the PR. However I have one concern on the mgd file. Can you please take a look ?

colemanw commented:
@monishdeb we'll have to rebase this again and move the upgrade code to 5.46. That'll be easier after merging #22224

monishdeb commented:
I have reviewed and marked #22224 with MoP. Reviewed the current PR in my local setup, everything works fine. Tested by submitting new Grant from and didn't encounter any issue

colemanw commented:
@monishdeb did you test all of these?

- Advanced Search
- Grants tab on contact summary screen
- Grant reports

colemanw commented:
@monishdeb I've got this rebased again to use the 5.46 upgrader. But I think search/reports still need testing per my above comment.

monishdeb commented:
@colemanw yep my bad ... I missed the advanced search, reports . Will update the PR shortly with search fixes.

colemanw commented:
Sounds good @monishdeb - FYI you'll have to reset your local branch because of the rebasing I just did.

seamuslee001 commented on ext/civigrant/CRM/Grant/DAO/Grant.php:
```diff
@@ -228,33 +228,33 @@ public static function &fields() {
           'table_name' => 'civicrm_grant',
           'entity' => 'Grant',
-          'bao' => 'CRM_Grant_BAO_Grant',
+          'bao' => 'CRM_Grant_DAO_Grant',
```
@colemanw any reason why this is being changed from the BAO to DAO?

colemanw replied:
:shrug: GenCode wanted to...

colemanw committed:
CiviGrant - Regenerate SQL

colemanw commented:
retest this please

eileenmcnaughton commented:
I chatted with @colemanw and as

1) we are early in the release cycle and
2) this PR has gotten stale a lot and
3) a lot of work has gone into it

I'm merging after a light 'click around' and we should continue to QA during the release cycle.

@colemanw told me there is still a report issue - although I didn't spot it when I ran the report just now (as demo user). However I did spot a couple of things

1) permissin incongruities - as demo user I could enable the extension but obvoulsy have a permission deficit as I could not
- click on the contact grants tab without error
- see civigrant menu items such as add & search
But I COULD
- see the grants tab that I couldn't click on 
- search for grants in advanced search
- access civi reports

2) demo site set up - once the above is sorted we should ensure civigrant is enabled & demo user has access on demo sites

FYI @seamuslee001 @demeritcowboy

MegaphoneJon commented:
Adding `CRM_Contact_BAO_Query_Interface {
::defaultReturnProperties()` breaks any Civi install that implements `hook_civicrm_queryObjects`.  I'm not sure there's anything to be done now except to submit PRs to extensions though.

eileenmcnaughton commented:
@MegaphoneJon I would be OK reverting - but I worry that extensions may have already adapted so that could break them....

Note it is implementing `CRM_Contact_BAO_Query_Interface` not the hook that is the issue (although probably both are done together). If I were maintaining an extension I'd probaly remove the extends though so it works on before & after
